libminifi/include/utils/StringUtils.h

```diff
@@ -92,172 +103,104 @@ class StringUtils {
     if (endString.size() > value.size())
       return false;
     return std::equal(endString.rbegin(), endString.rend(), value.rbegin());
   }
 
-  inline static std::string hex_ascii(const std::string& in) {
-    int len = in.length();
-    std::string newString;
-    for (int i = 0; i < len; i += 2) {
-      std::string sstr = in.substr(i, 2);
-      char chr = (char) (int) strtol(sstr.c_str(), 0x00, 16);
-      newString.push_back(chr);
-    }
-    return newString;
-  }
-
-  static std::string replaceMap(std::string source_string, const std::map<std::string, std::string> &replace_map) {
-    auto result_string = source_string;
-
-    std::vector<std::pair<size_t, std::pair<size_t, std::string>>> replacements;
-    for (const auto &replace_pair : replace_map) {
-      size_t replace_pos = 0;
-      while ((replace_pos = source_string.find(replace_pair.first, replace_pos)) != std::string::npos) {
-        replacements.emplace_back(std::make_pair(replace_pos, std::make_pair(replace_pair.first.length(), replace_pair.second)));
-        replace_pos += replace_pair.first.length();
+  std::string hex_ascii(const std::string& in);
+
+  template<class TChar, class U, typename std::enable_if<std::is_same<typename U::value_type, std::basic_string<TChar>>::value>::type* = nullptr>
+  std::basic_string<TChar> join(const std::basic_string<TChar>& separator, const U& container) {
+    typedef typename U::const_iterator ITtype;
+    ITtype it = container.cbegin();
+    std::basic_stringstream<TChar> sstream;
+    while(it != container.cend()) {
+      sstream << (*it);
+      ++it;
+      if(it != container.cend()) {
+        sstream << separator;
       }
     }
-
-    std::sort(replacements.begin(), replacements.end(), [](const std::pair<size_t, std::pair<size_t, std::string>> a,
-        const std::pair<size_t, std::pair<size_t, std::string>> &b) {
-      return a.first > b.first;
-    });
-
-    for (const auto &replacement : replacements) {
-      result_string = source_string.replace(replacement.first, replacement.second.first, replacement.second.second);
+    return sstream.str();
+  };
+
+  template<class TChar, class U, typename std::enable_if<std::is_same<typename U::value_type, std::basic_string<TChar>>::value>::type* = nullptr>
+  std::basic_string<TChar> join(const TChar* separator, const U& container) {
+    return join(std::basic_string<TChar>(separator), container);
+  };
+
+
+  template<class TChar, class U, typename std::enable_if<std::is_integral<typename U::value_type>::value>::type* = nullptr,
+      typename std::enable_if<!std::is_same<U, std::basic_string<TChar>>::value>::type* = nullptr>
+  std::basic_string<TChar> join(const std::basic_string<TChar>& separator, const U& container) {
+    typedef typename U::const_iterator ITtype;
+    ITtype it = container.cbegin();
+    std::basic_stringstream<TChar> sstream;
+    while(it != container.cend()) {
+      sstream << string_traits<TChar>::convert_to_string(*it);
```

Comment:
Some good comments, thanks. So I have a better understanding ( and now that I'm not watching a movie trying to review this from my phone ), why not allow the stringstream to do the conversion? It's, in my compiler impls, faster. I agree that to_string can be cleaner for simple or single conversions, but you are already using a stream. Other implementations of join I've seen for strings in c++ typically use iterators and then a stream and concat. What was the impetus for the inline conversion?

Reply:
For eg. uint8_t. 
Which is handled as char, so given the following:
```
std::list<uint_8t> l = { 10, 11 }
```
Using operator<< a new line and a tabulator would be printed instead of the numbers, while to_string properly converts them to string representation of the given number.

Reply:
This seems to enforce a semantic. I think basic stringstream assumes character if it is one. In that case, I don't see the need for the complexity as you could have one function that assumes what streamstream does and an overload that allows uint8_t to be interpreted with to_string. Two functions, much simpler and cleaner. Removing the rest of the changes will alleviate any linker issues.

Reply:
That could help removing the two template structs that provide convert_to_string methods, the price would be two different template specialisations (one for char and uint8_t, the other is wchar_t and uint8_t), so it can be done, but not sure if that makes it any simpler. 

The rest of the changes are coming from the fact that the this was a class, which cannot be extended with template functions. 
Changing to namespace means that the linkage changes, which naturally need to be verified on all platforms, but it cannot be avoided.

Reply:
Why would you need specializations for "*char" ? So that they can be dealt with as numerics? 

Can't have templates in classes???

Reply:
char: the specialisation is needed for string (aka std::basic_string<char>) and wstring (aka std::basic_string<wchar_t>). The latter one is especially needed as are to support Windows. 

Templates in classes: they are basically allowed, but full specialisation is not (https://en.cppreference.com/w/cpp/language/member_template), so it would fail any way: nor I can specialise the function for uint8_t, neither can add the current "convert_to_string" functions.

Reply:
I'm sorry I don't follow why that's absolutely necessary . We should have a discussion of what the differences are in a different format as I'm still -1 on this change even without the arm failures ( down to only two devices after an OS upgrade ).

Reply:
Reduced the impact by keeping the class (will add a follow-up), should be good now.